### PR TITLE
Merge in the desmos fork

### PR DESCRIPTION
**Pull request by jwmerrill**
Adds screen-reader support and other features, optimizations and fixes.

There are a couple large structural changes in here that touch pretty much every line in every file
* The codebase has been ported from JS to Typescript (see https://github.com/desmosinc/mathquill/pull/168 for just these changes)
* The codebase is now being formatted by prettier. There's a precommit hook to format staged files. (see https://github.com/desmosinc/mathquill/pull/182 for just these changes)

For review, it may be easiest to look at #947 for changes that landed before the TS port and the commits after 23d2a1b for changes that landed after the TS port.

In addition to the new options added in https://github.com/mathquill/mathquill/pull/947/, subsequent work added two new config options.
- `quietEmptyDelimiters: string;`
    - space-separated list of delimiters (such as parens, brackets, and curly braces) for which Mathquill will not display a gray box inside empty pairs. Desmos configures this as `quietEmptyDelimiters: '( ['` to avoid showing a gray box between empty parens and square brackets.
- `disableAutoSubstitutionInSubscripts: boolean;`
    - When `true`, `autoCommands` and `autoOperatorNames` are not substituted in subscripts. For example, if `pi` is an autocommand, typing `a` `_` `p` `i` would result in the subscript containing two letters instead of a single `π` symbol.

### CODE REVIEW

**Comment by jwmerrill** [comment]
Copying the commit comment from 9ac1090 for more visibility because I think it's important:

> Fixes https://github.com/mathquill/mathquill/issues/950 by reverting https://github.com/mathquill/mathquill/pull/764 and https://github.com/mathquill/mathquill/pull/779
> 
> Unfortunately, this leaves us back in a place where superscript and subscript placement isn't very good when they're attached to tall blocks, but I think if we're going to fix that, we need to find a way to do it that's more robust to continuing to edit the expression.

**Comment by sclower** [comment]
## New Config Options

Note: config.md has not been updated with all of the below options as Desmos uses some internally, and we may still find a need to change them before they are set in stone.

- ignoreNextMousedown: (_el:MouseEvent) => boolean;
- typingAsteriskWritesTimesSymbol?: boolean;
- typingSlashWritesDivisionSymbol: boolean;
- typingPercentWritesPercentOf?: boolean;
- resetCursorOnBlur?: boolean | undefined;
- enableDigitGrouping?: boolean; // adds visual spacing between digits
- disableCopyPaste?: boolean;
- onPaste?: () => void;
- onCut?: () => void;
- overrideTypedText?:(text:string) => void; // allows caller to intercept typed text before it is committed
- overrideKeystroke: (key:string, event:KeyboardEvent) => void; // allows caller to intercept keystroke before Mathquill's default handlers take over
- quietEmptyDelimiters: { [id:string]:any; }; // space-separated list of delimiters (such as parens, brackets, and curly braces) for which Mathquill will not provide a gray outline.
- disableAutoSubstitutionInSubscripts?: boolean; // allows caller to set whether auto commands, auto operators, and automatically parenthesized functions are expanded when typing into a subscript: e.g. should x_{pi} be changed to x_{\\pi}?

File: Makefile [hunk 1]
```diff
@@ -106,34 +114,44 @@ css: $(BUILD_CSS)
...
 clean:
 	rm -rf $(BUILD_DIR)
-
-$(PJS_SRC): $(NODE_MODULES_INSTALLED)
+# This adds an entry to your local .git/config file that looks like this:
+# [include]
+# 	path = ../.gitconfig
+# that tells git to include the additional configuration specified inside the .gitconfig file that's checked in here.
+setup-gitconfig:
+	@git config --local include.path ../.gitconfig
```
Comment (jwmerrill) [inline Makefile:122]: Note, running `make setup-gitconfig` once will make blame skip the commits where prettier was run, which helps a lot with preserving history.

We could consider documenting this somewhere more prominent, or trying to run this as part of install, or something like that.

**Comment by jaltekruse** [comment]
We're closing this PR in favor of creating a chain of PRs that segment the different parts of this work that will have cleaner diffs to look at.

**Pull request closed without merging**